keps/sig-cli/0015-kubectl-plugins.md

```diff
@@ -0,0 +1,152 @@
+It forces plugin scripts and executables to exist in a pre-determined location, requires basic information for a single 
+plugin to be declared explicitly in its own file, anddoes not provide a clear way to override existing command paths, 
+or provide additional subcommands without having to override a top-level command.
+
+A re-design plugins system allows us to implement extensibility requests from users that the current system cannot address
+(See https://github.com/kubernetes/kubernetes/issues/53640).
+
+### Goals
+
+* Support a simple plugin installation process (no additional config, users drop a binary in their `PATH`, for example, 
+  and they are then able to use them).
+  No additional configuration is needed, only the plugin executable.
+  A plugin's filename determines the plugin's intention, such as which path in the command tree it applies to:
+  `/usr/bin/kubectl-create-foobar` would, for example, override (or create) the command `kubectl create foobar`.
+  It is up to a plugin to parse any arguments and flags given to it. A plugin decides when an argument is a
+  subcommand, as well as any limitations or constraints that its flags should have.
+* Relay all information given to `kubectl` (via command line args) to plugins as-is.
+  Plugins receive all arguments and flags provided by users and are responsible for adjusting their behavior
+  accordingly.
+* Provide a way for plugins to override specific subcommands, while preventing other paths from being overriden.
+* Provide a standard package of common cli utilities that is consumed by `kubectl` and plugins alike.
+  These utilities would provide all of the major pieces needed to build a standard cli command, such as retrieving 
+  REST configuration, REST clients, printing, etc. (See https://github.com/kubernetes/features/issues/581).
+
+### Non-Goals
+
+* The new plugin mechanism will not be a "plugin installer" or wizard. It will not have specific or baked-in knowledge 
+  regarding a plugin's location or composition, nor will it it provide a way to download or unpack plugins in a correct 
+  location.
+* Plugin discovery is not a main focus of this mechanism. As such, it will not attempt to collect data about every 
+  plugin that exists in an environment.
+* Plugin management is out of the scope of this design. A mechanism for updating and managing lifecycle of existing 
+  plugins should be covered as a separate design (See https://github.com/kubernetes/community/pull/2340).
+
+## Proposal
+
+### Scenarios
+
+* Developer wants to create and expose a plugin to `kubectl`.
+  They use a programming language of their choice and create an executable file.
+  The executable's filename consists of the command path to replace, and is prefixed with `kubectl-`.
+  The executable file is placed on the user's PATH, preferably under `/usr/local/bin`.
+  If a plugin is written using a dynamic language, and depends on logic defined in additional files, those
```

Comment:
Why put this here?  Seems like extraneous detail.

Reply:
I can move this to the "Implementation Details" section if it makes more sense there. I wanted to clarify how certain aspects of making a complex plugin, for example, would be impacted by this design